topology/probes/ovs_of.go

```diff
@@ -0,0 +1,503 @@
+	g.Link(bridgeNode, ruleNode, graph.Metadata{"RelationType": "ownership"})
+}
+
+// delRule deletes a rule from the the graph.
+func delRule(rule *Rule, probe *BridgeOfProbe) {
+	g := probe.OvsOfProbe.Graph
+	g.Lock()
+	defer g.Unlock()
+
+	ruleNode := g.LookupFirstNode(graph.Metadata{"UUID": rule.UUID})
+	if ruleNode != nil {
+		g.DelNode(ruleNode)
+	}
+}
+
+// monitor monitors the openflow rules of a bridge by launching a goroutine. The context is used to control the execution of the routine.
+func monitor(ctx context.Context, probe *BridgeOfProbe) error {
+	ofp := probe.OvsOfProbe
+	command, err1 := ofp.makeCommand([]string{"ovs-ofctl", "monitor"}, probe.Bridge, "watch:")
+	if err1 != nil {
+		return err1
+	}
+	lines, err := LaunchContinuousOnSwitch(ctx, command)
+	if err != nil {
+		return err
+	}
+	go func() {
+		prefix := probe.Host + "-" + probe.Bridge + "-"
+		for line := range lines {
+			event, err := ParseEvent(line, probe.Bridge, prefix)
+			if err == nil {
+				switch event.Action {
+				case "ADDED":
+					err = completeRule(ofp, &event)
+					if err != nil {
+						logging.GetLogger().Error(err.Error())
+					}
+					rule := event.Rule
+					_, here := probe.Rules[rule.UUID]
+					if !here {
+						probe.Rules[rule.UUID] = rule
+						addRule(rule, probe)
+					}
+				case "DELETED":
+					delete(probe.Rules, event.Rule.UUID)
+					delRule(event.Rule, probe)
+				}
+			} else {
+				logging.GetLogger().Errorf("Error while monitoring %s@%s: %s", probe.Bridge, probe.Host, err.Error())
+			}
+		}
+
+	}()
+	return nil
+}
+
+// NewBridgeProbe creates a probe and launch the active process
+func (o *OvsOfProbe) NewBridgeProbe(host string, bridge string, uuid string) (*BridgeOfProbe, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	address, ok := o.Translation[bridge]
+	if !ok {
+		address = bridge
+	}
+	probe := BridgeOfProbe{
+		Host:       host,
+		Bridge:     bridge,
+		UUID:       uuid,
+		Address:    address,
+		BridgeNode: nil,
+		OvsOfProbe: o,
+		Rules:      make(map[string]*Rule),
+		cancel:     cancel}
+	err := monitor(ctx, &probe)
+	return &probe, err
+}
+
+func (o *OvsOfProbe) makeCommand(commands []string, bridge string, args ...string) ([]string, error) {
+	commandLine := []string{}
+	if o.Sudo {
+		commandLine = []string{"sudo"}
+	}
+	commandLine = append(commandLine, commands...)
+	if strings.HasPrefix(bridge, "ssl:") {
+		if o.sslOk {
+			commandLine = append(commandLine,
+				bridge,
+				"--certificate", o.Certificate,
+				"--ca-cert", o.CA, "--private-key", o.PrivateKey)
+		} else {
+			return commands, errors.New("Certificate, CA and private keys are necessary for communication with switch over SSL")
+		}
+	} else {
+		commandLine = append(commandLine, bridge)
+	}
+	commandLine = append(commandLine, args...)
+	return commandLine, nil
+}
+
+// OnOvsBridgeAdd is called when a bridge is added
+func (o *OvsOfProbe) OnOvsBridgeAdd(uuid string, row *libovsdb.RowUpdate) {
```

Comment:
What is the gain ? The probe is really a subprobe. It cannot work without the information supplied on the topology by the ovsdb probe. A regular listener will just be slower because it will listen to all node events. At least, we are sure that the initial configuration of the bridge has been analyzed.